### DISCUSSION TITLE
How do I tell the compiler that a type supports a specific interface?

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#1840] [Language Ideas]
I want to write the IndexOf method for `Span` and type `T` without limits: `public static int IndexOfEquals<T>(this Span<T> span, T value)`.

You can notice that if the type `T` implements the interface `IEquatable<T>` then for such `T` corresponding method is already exists and we just need to call it, but the compiler will not allow us to do that.

Next code have incorrect casting `(T: IEquatable<T>)value`. How to make it correct?
```cs
        public static int IndexOfEquals<T>(this Span<T> span, T value)
        {
            if (typeof(IEquatable<T>).IsAssignableFrom(typeof(T)))
                return System.MemoryExtensions.IndexOf(span, (T: IEquatable<T>)value);
            // something else
        }
```

How can I call `System.MemoryExtensions.IndexOf` method in case if `T` type supports `IEquatable<T>` interface?

**Comment by theunrepentantgeek** [comment]
What's wrong with this?

```
public static int IndexOfEquals<T>(this TimeSpan<T> span, T value)
{
    if (value is IEquatable<T> e)
    {
        return System.MemoryExtensions.IndexOf(span, e);
    }

    return 42;
}
```

**Comment by AlexRadch** [comment]
@theunrepentantgeek it is wrong because we have `Span<T> span` but not `Span<IEquatable<T>> span`. It was first that I tried.

**Comment by jnm2** [comment]
@AlexRadch I don't understand; the second parameter to `MemoryExtensions.IndexOf` must be `T` or `ReadOnlySpan<T>`, not `IEquatable<T>`. What's the larger purpose of your `IndexOfEquals`?

**Comment by svick** [comment]
@jnm2 The second parameter (of the overload of interest here) must be `T where T : IEquatable<T>`. As I understand it, the question here is: How can I call that method if I have unconstrained `T` (after somehow dynamically checking that `T` implements `IEquatable<T>`)?

**Comment by jnm2** [comment]
Ah, similar to https://github.com/dotnet/csharplang/issues/1315?

**Comment by svick** [comment]
@jnm2 Possibly, though that seems to be geared towards specializing for a list of specific types, not for all types that satisfy a constraint.

**Comment by svick** [comment]
@AlexRadch I think the usual answer to questions of the form "The C# type system doesn't let me do this, what do I do?" is "Bypass it by using reflection". The issue is that you can't use reflection with `Span<T>`, but you can also work around that. Though the work around I came up with is quite convoluted:

```c#
// Note: ignore the class names, couldn't think of better ones
abstract class ClosureBase<T>
{
    public static ClosureBase<T> Instance = (ClosureBase<T>)Activator.CreateInstance(
        typeof(Closure<>).MakeGenericType(typeof(T)));

    public abstract int IndexOf(ReadOnlySpan<T> span, T value);
}

sealed class Closure<T> : ClosureBase<T> where T : IEquatable<T>
{
    public override int IndexOf(ReadOnlySpan<T> span, T value) =>
    	MemoryExtensions.IndexOf(span, value);
}    
    
public static int IndexOfEquals<T>(this Span<T> span, T value)
{
    if (typeof(IEquatable<T>).IsAssignableFrom(typeof(T)))
    {
        return ClosureBase<T>.Instance.IndexOf(span, value);
    }
    return 42;
}
```

**Comment by jnm2** [comment]
If there was a way to reinterpret the `Span<T>` as a `Span<EqualityWrapper<T>>`, you could move the equality logic into `EqualityWrapper<T>.Equals(T)`.

I'm new to the APIs and I can't figure out how to reinterpret the span memory this way, though. It should be safe since the size and layout of the wrapper struct should be identical to the bits in the `T` span.

**Comment by jnm2** [comment]
Wait, think I got it. This is passing my tests:

```cs
public static unsafe int IndexOfEquals<T>(this Span<T> span, T value)
{
    fixed (void* pointer = &Unsafe.As<T, byte>(ref span.GetPinnableReference()))
    {
        return System.MemoryExtensions.IndexOf(
            MemoryMarshal.CreateSpan(ref Unsafe.AsRef<EqualityWrapper<T>>(pointer), span.Length),
            new EqualityWrapper<T>(value));
    }
}

private readonly struct EqualityWrapper<T> : IEquatable<EqualityWrapper<T>>
{
    private readonly T value;

    public EqualityWrapper(T value)
    {
        this.value = value;
    }

    public bool Equals(EqualityWrapper<T> other)
    {
        // Whatever you want here—this is what I’d probably go for.
        return EqualityComparer<T>.Default.Equals(value, other.value);
    }

    public override bool Equals(object obj)
    {
        return obj is EqualityWrapper<T> other && Equals(other);
    }

    public override int GetHashCode()
    {
        return value == null ? 0 : value.GetHashCode();
    }
}
```

**Comment by jnm2** [comment]
The above also handles searches for `null` properly, whereas `System.MemoryExtensions.IndexOf` throws!

**Comment by AlexRadch** [comment]
@jnm2 Thanks for your very cool code!

**Comment by AlexRadch** [comment]
@svick Thanks for your very cool code! I think it is more tricky code than @jnm2 code but more fast.
@jnm2 code will remove optimizations for standard types like `string` or `int`.

**Comment by jnm2** [comment]
@jkotas Could you please check that [my code](https://github.com/dotnet/csharplang/issues/1840#issuecomment-419466894) is safe, interpreting contiguous `T` bits as contiguous `WrapperStruct<T>` bits?

**Comment by jnm2** [comment]
@AlexRadch 

> @jnm2 code will remove optimizations for standard types like string or int.

I don't think this is the case. `EqualityComparer<T>.Default.Equals` is a JIT instrinsic and compiles to the same code as comparing ints with `==` when `T` is `int`.

Compare: [generic](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wKlIrwGUALRAJwAcAZRYAOgCUArgDs4AS1QBTWuQDaAKVFwA4hKESWogMYAKOAE82EgPYAzbaJEBKSwF1yLCYgAmRoQBs9uZHBYDNcXABRAEcBRDdFPQB1FkQ2QxYAHgAVAD5cEFwASRCwuG43CUTc8MiYuISU1NTyAG9yXEbcNg0AN0Q4CVwHZ1cPXGTcdrcBKXIGpswAZiDQ0v1y+PVtQeHRywnG+rIm3dw4RlFkHjWugF4h8NHaXYBfTdwH6dxgIyM3WbC3ZG0SiIXYkskmlcEYDuoNjsmts9k0APRw3BRZidVrqXB6IwCXAAd0QIlwjHUEkAKAQHI64Ck4lHZQCYBE5miwjMACp4AOZGXAmIwsHgPXaYADsn3megAwkZUGxWOoqjwACISEyIARuOA8P4/U4wUHg3mnSw3Jr3MhPGZGNEsDROLqvd4i77aZkAKwk/lBwGdkN2MNh+GFLspyAdZUBlRBYKJLFwADIYw6fpGIUbGiazaDLdauhYAio4AAJFCMCU27Te6H8ybC064M4XISqj4AflweEypx4ecLyGLRlLhoeJpNQA===) versus [handwritten specialized `int`](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wKlLPICcBTRAEwHsA7AGwE9dk49AK4BjOLgCiARyGIuASzg8A6vUQAHdY3q4QuAJLTZcRMC6MAPEbmKVazdoB85AN7lcH3OvryAbojhGXCZWTl5ceQ5xfy4hRlp3T0wAZkkZGyVVDS16AApI6Lk4gEpEjzcyTyrcOAALeWQAOhi43ABeXBb4stwAXx6elNxgNjYuNNkuZFzrBUz7HNw2Ou1Sys8K6s8Aem3cZVqAxl9tXB42IVwAd0Qo3FrtRkAUAjqGiORrw/F9QEwCFi96GxgKZwgBzNi4ABmbHojR6VUwAHYJhkeABhNiodSIJj0CwFRyNAAijEhiCEXDgjVm0y6MCWK1hXWKtCq/To6w8QzYJ3oPhYQRGYxRU1yQIAVowxEtgOK1lVNlt8MiJe8RbYsg4dMsHjoAGR6kXTHWrVmedmDVI87T8oIFXAAcUYcAAEihahiBbl5Rt4UlkV0zR52eygA); the asm is the same.

**Comment by AlexRadch** [comment]
@jnm2 I think that for some standard types like `char` or `byte`  method `IndexOf<T>` does not use `IEquatable<T>` interface at all for speed. But it will use this interface for `EqualityWrapper<char>` type.

**Comment by jkotas** [comment]
> @jkotas Could you please check that my code is safe

Yes, it should work fine, but I think it is unnecessary ceremony. You can just have a simple loop (maybe unrolled) to achieve the same effect. You are not gaining much by bending backwards to call `System.MemoryExtensions.IndexOf`.

> some standard types like char or byte method IndexOf<T> does not use IEquatable<T> interface

Correct. These optimizations won't kick in for the above.

**Comment by jnm2** [comment]
@AlexRadch Ah, yes I see.

@jkotas Thank you! Good to know that.

**Comment by mikedn** [comment]
Seem like the "reinterpret" acrobatics can be reduced using `MemoryMarshal.Cast`
```C#
Span<Foo> s = new Foo[42];
Span<EqualityWrapper<Foo>> e = MemoryMarshal.Cast<Foo, EqualityWrapper<Foo>>(s);
```

**Comment by jnm2** [comment]
@mikedn We need to use it with unconstrained generics. That fails at both compile time (`TTo` is `where TTo : struct`) and runtime:

```cs
        public static Span<TTo> Cast<TFrom, TTo>(Span<TFrom> span)
            where TFrom : struct
            where TTo : struct
        {
            if (RuntimeHelpers.IsReferenceOrContainsReferences<TFrom>())
                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(TFrom));
            if (RuntimeHelpers.IsReferenceOrContainsReferences<TTo>())
                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(TTo));
```

**Comment by jnm2** [comment]
@jkotas 

> Correct. These optimizations won't kick in for the above.

Actually, what is preventing the `EqualityWrapper<int>.Equals` call from being inlined?

**Comment by mikedn** [comment]
> We need to use it with unconstrained generics. That fails at both compile time (TTo is where TTo : struct) and runtime

Ah, right, that's unfortunate. One can anyway achieve the same thing without using `Cast` so the restriction seems a bit pointless.

**Comment by tannergooding** [comment]
There should probably just be a `MemoryExtensions.IndexOf(Span<T>, T, IEqualityComparer<T>)` overload, to cover the case where `T` doesn't implement `IEquatable<T>`

**Comment by jkotas** [comment]
> Actually, what is preventing the EqualityWrapper<int>.Equals call from being inlined?

Nothing fundamental. Inlining is not the optimization dropped here. The optimization dropped here is the manual vectorization for byte, etc.

**Comment by jnm2** [comment]
These are solutions to @AlexRadch's specific problem, but the general problem could be solved with https://github.com/dotnet/csharplang/issues/399. Each interface check could cause the code inside the succeeding branch to intersect the type (`T & IEnumerable<T>` etc). Pretty harrowing to implement without CLR support, though, since the compiler would have to do the same reflection workaround for generic constraints that @svick posted at the top.

@jkotas It would be pretty cool if it could be both inlined and vectorized.

**Comment by AlexRadch** [comment]
I created proposal to add pattern matching for types https://github.com/dotnet/csharplang/issues/1839. Also my proposal mostly is duplicate of  https://github.com/dotnet/csharplang/issues/631 proposal.
So I think pattern matching for types should be added to c# in future.
